keps/sig-api-machinery/2144-clientgo-apply/README.md

```diff
@@ -0,0 +1,578 @@
+  Use the existing go structs to construct an apply configuration, serialize
+  it to JSON, and pass it to `Patch`. This can cause zero valued required
+  fields being accientally included in the apply configuration resulting
+  in fields being accidentally set to incorrect values and/or fields accidentally
+  being accidentally being clamed as owned.
+
+Both sig-api-machinery and wg-api-expression agree that this enhancement is
+required for server side apply to be promoted to GA.
+
+### Goals
+
+Introduce a typesafe, programmatic way to call server side apply using the typed
+client in client-go.
+
+Express all apply configurations in Go that can be expressed in
+YAML. Specifically, an apply request must include only fields that are set by
+the applier and exclude those not set by the applier.
+
+Validate this enhancement meets the needs of developers:
+
+- An developer not directly involved in this enhancement successfully converts
+  a 1st party controller (one in github.com/kubernetes/kubernetes) to use this
+  enhancement.
+- A representative group of the developer community is made aware of this
+  proposed enhancement, is given early access to it via a fork of
+  controller-tools with the requisite generators, and is given the opportunity
+  to try it out and provide feedback.
+
+### Non-Goals
+
+Enhancements to client-go's dynamic client. The client-go dynamic client already
+supports Apply via Patch, which is adequate for the dynamic client.
+
+Protobuf support. Apply does not support protobuf, and it will not be added with
+this enhancement.
+
+## Proposal
+
+`Apply` functions should be included in the typed clients generated for
+client-go and should accept the apply configuration using a strongly typed
+representation, which will need to be generated for this purpose.
+
+### Risks and Mitigations
+
+#### Poor adoption
+
+Risk: Developers adoption is poor, either because the aesthetics/ergonomics are
+not to their liking or the functinality is insufficient to do what they need to
+do with it. This could lead to (a) poor server side apply adoption, and/or (b)
+developers building alternate solutions.
+
+Mitigation: We are working with the kubebuilder community to
+get hands on feedback from developers to guide our design decisions around
+aesthetics/ergonomics with a goal of having both client-go and kubebuilder
+take an aligned approach to adding apply to clients in a typesafe way.
+
+## Design Details
+
+### Apply functions
+
+The client-go typed clients will be extended to include Apply functions, e.g.:
+
+```go
+func (c *deployments) Apply(ctx Context, fieldManager string, deployment *appsv1apply.Deployment, opts metav1.ApplyOptions, subresources ...string) (*Deployment, error)
+```
+
+`ApplyOptions` will be added to metav1 even though `PatchOptions` will continue
+to be used over the wire. This will make it obvious in the Apply function
+signature that `fieldManager` is required.
+
+```go
+type ApplyOptions struct {
```

Comment:
you made me look :)  PatchOptions has a FieldManager today.  Why isn't that reflected here?

Reply:
It's deliberate. The idea here is to make fieldManger required instead of optional. The way we're proposing doing that is by removing it from the Options and adding it to the Apply function as a parameter.

This is related to the whole "should field manager have a default" question.